save-core/src/commonMain/kotlin/org/cqfn/save/core/files/ConfigDetector.kt

```diff
@@ -0,0 +1,79 @@
+import okio.Path
+
+/**
+ * A class that is capable of discovering config files hierarchy.
+ */
+class ConfigDetector {
+    /**
+     * Try to create SAVE config file from [file].
+     *
+     * @param file a [Path] from which SAVE config file should be built.
+     * @return [TestSuiteConfig] or null if no suitable config file has been found.
+     */
+    fun configFromFile(file: Path): TestSuiteConfig? = discoverConfigWithParents(file)
+        ?.also { config ->
+            // fill children for parent configs
+            config.parentConfigs(wihSelf = true).toList().reversed()
+                .zipWithNext().forEach { (parent, child) ->
+                    parent.childConfigs.add(child)
+                }
+            // discover all descendant configs of [config]
+            val locationsFlattened = config.location.parent!!.findAllFilesMatching { it.isDefaultSaveConfig() }.flatten()
+            val configs = mutableListOf(config)
+            locationsFlattened
+                .drop(1)  // because [config] will be discovered too
+                .forEachIndexed { index, path ->
+                    configs.add(
+                        TestSuiteConfig(
+                            "todo: read from file",
+                            "todo: read from file",
+                            path,
+                            configs.find { discoveredConfig ->
+                                discoveredConfig.location ==
+                                        locationsFlattened.take(index + 1).reversed().find { it.parent in path.parents() }!!
+                            }!!
+                        ).also {
+                            it.parentConfig!!.childConfigs.add(it)
+                        }
+                    )
+                }
+        }
+
+    private fun discoverConfigWithParents(file: Path): TestSuiteConfig? = when {
+        // if provided file is a directory, try to find save.toml inside it
+        FileSystem.SYSTEM.metadata(file).isDirectory -> file
+            .findChildByOrNull { it.isDefaultSaveConfig() }
+            ?.let { discoverConfigWithParents(it) }
+        // if provided file is an individual test file, we search a config file in this and parent directories
+        file.name.matches(testResourceFilePattern) -> file.parents()
+            .mapNotNull { dir ->
+                dir.findChildByOrNull { it.isDefaultSaveConfig() }
+            }
+            .firstOrNull()
+            ?.let { discoverConfigWithParents(it) }
+        // if provided file is save.toml, create config from it
+        file.isDefaultSaveConfig() -> testSuiteConfigFromFile(file)
+        else -> null
+    }
+
+    private fun testSuiteConfigFromFile(file: Path): TestSuiteConfig = TestSuiteConfig(
+        "todo: read from file",
+        "todo: read from file",
+        file,
+        file.parents()
+            .drop(1)  // because immediate parent already contains [this] config
+            .mapNotNull { parentDir ->
```

Comment:
so you do mapNotNull, but in this map can return null in findChildByNull

Reply:
Yes, this way we can skip parent directories that don't contain a config file and find it on higher levels